plugins/CoreVisualizations/Visualizations/Sparklines/SparklinePeriodSelector.php

```diff
@@ -0,0 +1,123 @@
+                $params['compareDates'][] = $period->getRangeString();
+                $params['comparePeriods'][] = $params['period']; // need to ensure to use same period for both the base and the comparison period
+            }
+        } else {
+            // when not comparing we select the last 30 days/weeks/months/years.
+            // For period range it will select the selected range with a day metric
+            $params = $this->config->getGraphParamsModified($params);
+            if ($originalPeriod->getLabel() === 'range') {
+                // when a longer range is selected, then we select a higher period for improved performance and also to see trends better
+                // eg when selecting a range of 2 years then we rather show 24 months instead of 730 day points.
+                $params['period'] = $highestPeriodInCommon;
+                $params['date'] = $originalPeriod->getRangeString();
+            }
+        }
+
+        return $params;
+    }
+
+    public function getComparisonPeriodObjects($comparePeriods, $compareDates): array
+    {
+        $periods = [];
+        if (!empty($comparePeriods)) {
+            foreach ($comparePeriods as $periodIndex => $period) {
+                if ($periodIndex === 0) {
+                    continue; // this is the original period and we need to skip it
+                }
+                $date = $compareDates[$periodIndex];
+                $periods[] = Factory::build($period, $date);
+            }
+        }
+        return $periods;
+    }
+
+    /**
+     * Given all the periods, determine what is the lowest period we can use. For example if someone compares
+     * a range of 2 years with a range of 2 days, then we still need to use "day" periods in the sparkline as otherwise
+     * you would not be able to compare the 2 different sparklines when they use different periods.
+     *
+     * For best performance and for better detecting trends we try to select the highest period if all the periods given
+     * allow it.
+     *
+     * @param Period $originalPeriod
+     * @param $comparisonPeriods
+     * @return string
+     */
+    public function getHighestPeriodInCommon(Period $originalPeriod, $comparisonPeriods): string
+    {
+        $lowestNumDaysInRange = $this->getNumDaysDifference($originalPeriod->getDateStart(), $originalPeriod->getDateEnd());
+
+        if (!empty($comparisonPeriods)) {
+            foreach ($comparisonPeriods as $period) {
+                $numDaysInRange = $this->getNumDaysDifference($period->getDateStart(), $period->getDateEnd());
+                if ($numDaysInRange < $lowestNumDaysInRange) {
+                    $lowestNumDaysInRange = $numDaysInRange;
+                }
+            }
+        }
+
+        $periodToUse = 'day';
+        if ($lowestNumDaysInRange >= 730) {
+            $periodToUse = 'month'; // rather than fetching > 730 day periods we prefer fetching 24+ months and shows trends better
```

Comment:
Would it be useful to have these values as configurable settings?

Reply:
Thanks for checking @bx80 . Not needed in this case. We'd only add configs if really needed. Should a use case come up at some point then we can always add it later. 👍